Create .deb release CI

maddsua opened the pull request:
Closes #80 

Add a CI pipeline that builds .deb package files.

This allows the package to be installed to older systems such as Debian 11 Bullseye, and everything more recent.

It's also possible to include more platforms by utilizing the matrix strategy.

Dependencies:
- libogg 1.3.4
- libc 2.31

---

Example CI run: https://github.com/maddsua/opustags/actions/runs/14940009362
Example release: https://github.com/maddsua/opustags/releases/tag/1.10.1

maddsua committed:
create .deb package release ci

fmang commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,122 @@
+  build-release:
+    needs: ["create-release"]
+    runs-on: ubuntu-latest
+    container: gcc:11.5.0-bullseye
```
Didn’t we conclude jammy would be the better choice?

Plus, if you wrote `runs-on: ubuntu-22.04`, the containerization would not be necessary, making the action more lightweight.

fmang commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,122 @@
+          GH_TOKEN: ${{ github.token }}
+          GH_REPO: ${{ github.repository }}
+
+  publish-release:
```
If this job runs automatically I don’t want it. I publish releases manually after I’ve reviewed them. Same for upload-assets. I don’t publish releases often and prefer to be careful.

fmang commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,122 @@
+      - "v*.*.*"
+
+jobs:
+  create-release:
```
Is this job useful? By making build-release depend on it I won’t be able to build or re-build the .deb file for an existing release, nor be able to test the workflow on a separate branch.

fmang commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,122 @@
+      - name: Build debian binary
+        env:
+          CXX: g++
+          CXXFLAGS: -DCMAKE_BUILD_TYPE=Release -D_FORTIFY_SOURCE=3 -D_GLIBCXX_ASSERTIONS -O2 -flto=auto -Wall -Wextra -Werror -fstack-protector-strong -fstack-clash-protection -fcf-protection
```
Where did you find these?

Passing -DCMAKE_BUILD_TYPE here looks erroneous. This is supposed to be a CMake flag, not a CXXFLAGS.

-Werror is good during development, but is not appropriate after I’ve pushed the tag. If I can’t at least run the workflow in a branch this is going to cause a lot of noise.

fmang commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,122 @@
+          mkdir -p debian/usr/bin
+          mkdir -p debian/DEBIAN
+          cp target/opustags debian/usr/bin/
+          chmod 775 debian/usr/bin/opustags
```
Please follow the installation procedure in the README. Including the binary is not enough.

maddsua replied:
Yup my bad, will adjust the action a bit later

fmang commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,122 @@
+          echo "Version: ${{ needs.create-release.outputs.pkg_version }}" >> ${CONTROL_FILE}
+          echo "Depends: libogg0 (>= 1.3.4) | libogg (>= 1.3.4)" >> ${CONTROL_FILE}
+          echo "Priority: optional" >> ${CONTROL_FILE}
+          echo "Description: Ogg Opus tags editor " >> ${CONTROL_FILE}
```
A here-doc would be more readable and more efficient.

```sh
cat <<EOF > debian/DEBIAN/control
Package: opustags
Architecture: amd64
…
EOF
```

maddsua commented:
Addresses your concerns, namely:

1. The pipeline is now triggered by a release publish event. You create a release yourself - and it builds and uploads it's respective assets. Version pattern matching is preserved just in case you decide to change the tag schema, so that it doesn't break dpkg.
2. Using your modified build instructions now. I don't think the exact instructions in readme look right. CMake ends up being run in an empty dir, and it's not happy about it.
3. Yup, using here-doc now

maddsua commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,83 @@
+      - name: CMake build
+        run: |
+          mkdir build
+          cmake -DCMAKE_INSTALL_PREFIX=./build/usr -S .
```
Okay, sorry but I really don't see the problem here. That does copy the needed files to an appropriate location, so why does that even matter? Yeah, there are the conventions, but using one or another doesn't really change the outcome does it?

fmang replied:
The prefix tells the executable where to find its resource files once installed, including translations or documentation, and is more generally used everywhere a path is hard-coded. It’s true opustags does not use it today, but the day it does the prefix you suggest is guaranteed to break the program. `/usr` is the most meaningful prefix here.

maddsua replied:
Alright, so if I understand that correctly, it must be `... -DCMAKE_INSTALL_PREFIX=/usr ...` followed by `make install DESTDIR=/build/usr`?

maddsua committed:
fix release ci install path

maddsua commented:
Just updated the path as previously discussed. Seems to behave as expected.

fmang commented:
Looks good to me. Thank you for your contribution and your patience.

I’d like to do some maintenance in the code base and release 1.11, but I can’t predict when this is going to happen exactly.

maddsua commented:
Glad that we've been able to bring this one to completion!
